sources/imdsclient/src/lib.rs

```diff
@@ -0,0 +1,228 @@
+/*!
+The imdsclient library provides high-level methods to interact with the AWS Instance Metadata Service.
+See the documentation for submodules [`fetch_dynamic`], [`fetch_metadata`], and [`fetch_userdata`] for high-level helpers.
+
+For more control, and to query IMDS without high-level wrappers, there is also a [`fetch_imds`] method.
+This method is useful for specifying things like a pinned date for the IMDS schema version.
+*/
+
+#![deny(rust_2018_idioms)]
+
+use http::StatusCode;
+use log::{debug, info, trace};
+use reqwest::blocking::Client;
+use snafu::ResultExt;
+
+const IMDS_BASE_URI: &str = "http://169.254.169.254";
+const IMDS_PINNED_DATE: &str = "2020-10-27";
+
+// Currently only able to get fetch session tokens from `latest`
+const IMDS_SESSION_TOKEN_URI: &str = "http://169.254.169.254/latest/api/token";
+
+/// A client for making IMDSv2 queries.
+/// It obtains a session token when it is first instantiated and is reused between helper functions.
+pub struct ImdsClient {
+    client: Client,
+    session_token: String,
```

Comment:
I'm a little wary about storing the token without giving the user the option to specify how long it will live.  A user may try to use the client in a process that lasts longer than 60 seconds.

I think a default value of 60 seconds is fine as it covers most of the use cases we have (short quick programs that make a request or three quickly), but I'd like to have an additional `new..`  and/or `fetch...` method that gives the user the opportunity to define the token TTL.

In Rust-land, this could be implemented as an additional method like `ImdsClient::new_with_token_ttl()`.  Under the covers it could call the `fetch_token_with_ttl()` method.  Doing it this way avoids `Option`s in the method signature, which requires a user to pass a `Some` or `None` (which IMHO in this case is not a very friendly API).

Reply:
Or should we implement automatic token refresh?

Reply:
^ also a good option. :)